monitoring/v1/types.go

```diff
@@ -0,0 +1,362 @@
+	// evaluated at the current time, and all resultant time series become pending
+	// or firing alerts.  This is most often a string representing a PromQL
+	// expression, e.g.: mapi_current_pending_csr > mapi_max_pending_csr
+	// In rare cases this could be a simple integer, e.g. a simple "1" if the
+	// intent is to create an alert that is always firing.  This is sometimes used
+	// to create an always-firing "Watchdog" alert in order to ensure the alerting
+	// pipeline is functional.
+	//
+	// +kubebuilder:validation:Required
+	Expr intstr.IntOrString `json:"expr"`
+
+	// for is the time period after which alerts are considered firing after first
+	// returning results.  Alerts which have not yet fired for long enough are
+	// considered pending.
+	//
+	// +optional
+	For Duration `json:"for,omitempty"`
+
+	// labels to add or overwrite for each alert.  The results of the PromQL
+	// expression for the alert will result in an existing set of labels for the
+	// alert, after evaluating the expression, for any label specified here with
+	// the same name as a label in that set, the label here wins and overwrites
+	// the previous value.  These should typically be short identifying values
+	// that may be useful to query against.  A common example is the alert
+	// severity, where one sets `severity: warning` under the `labels` key:
+	//
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// annotations to add to each alert.  These are values that can be used to
+	// store longer additional information that you won't query on, such as alert
+	// descriptions or runbook links.
+	//
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+// AlertingRuleStatus is the status of an AlertingRule resource.
+type AlertingRuleStatus struct {
+	// observedGeneration is the last generation change you've dealt with.
+	//
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// prometheusRule is the generated PrometheusRule for this AlertingRule.  Each
+	// AlertingRule instance results in a generated PrometheusRule object in the
+	// same namespace, which is always the openshift-monitoring namespace.
+	//
+	// +optional
+	PrometheusRule PrometheusRuleRef `json:"prometheusRule,omitempty"`
+}
+
+// PrometheusRuleRef is a reference to an existing PrometheusRule object.  Each
+// AlertingRule instance results in a generated PrometheusRule object in the same
+// namespace, which is always the openshift-monitoring namespace.  This is used to
+// point to the generated PrometheusRule object in the AlertingRule status.
+type PrometheusRuleRef struct {
+	// This is a struct so that we can support future expansion of fields within
+	// the reference should we ever need to.
+
+	// name of the referenced PrometheusRule.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=2048
+	Name string `json:"name"`
```

Comment:
What's the valid character set for a name?

Reply:
Any valid Go string.